### PR TITLE
[BCFR-203] Improve CR value comparator querying (topics and data words) by doing encoding in the relayer

### PR DESCRIPTION
**Pull request by ilija42**
### Description

- Implement QueryKey Value Comparator filters encoding. This enables the caller to just provide a raw value for filtering which then gets encoded and modified in the relayer to match the on chain version. For eg. chain agnostic code had to hash a string value like solc to filter over topic values, but now can just provide a raw string value which will be encoded in the EVM Chain Reader impl.  This works for filtering over indexed topics and simpler cases of non-indexed topics aka data words where offsets don't have to be calcualted.

### Requires Dependencies
[common](https://github.com/smartcontractkit/chainlink-common/pull/689)

### CODE REVIEW

**Review by EasterTheBunny — DISMISSED** [review]

File: core/services/relay/evm/chain_reader.go [hunk 1]
```diff
@@ -269,59 +258,117 @@ func (cr *chainReader) addEvent(contractName, eventName string, a abi.ABI, chain
...
+			if err := cr.addEncoderDef(contractName, dwTypeID, abi.Arguments{abi.Argument{Type: dWDetail.Type}}, nil, nil); err != nil {
+				return nil, nil, fmt.Errorf("%w: failed to init codec for data word %s on index %d querying for event: %q", err, genericName, dWDetail.Index, eventName)
+			}
+
+			dwCodecTypeID := codec.WrapItemType(contractName, dwTypeID, true)
+			dwsCodecTypeInfo[dwCodecTypeID] = cr.parsed.EncoderDefs[dwCodecTypeID]
 
-	// TODO can this be simplified? Isn't this same as inputInfo.Modifier()? BCI-3909
-	inMod, err := def.InputModifications.ToModifier(codec.DecoderHooks...)
+			dWsDetail[genericName] = dWDetail
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go:343]: Just realised that this [PR](https://github.com/smartcontractkit/chainlink/pull/13992) ran over this [PR](https://github.com/smartcontractkit/chainlink/pull/14096), so this TODO is getting re-resolved

File: core/capabilities/ccip/configs/evm/contract_reader.go [hunk 2]
```diff
@@ -140,11 +140,6 @@ var SourceReaderConfig = evmrelaytypes.ChainReaderConfig{
 				consts.EventNameCCIPMessageSent: {
 					ChainSpecificName: mustGetEventName("CCIPMessageSent", onrampABI),
 					ReadType:          evmrelaytypes.Event,
-					EventDefinitions: &evmrelaytypes.EventDefinitions{
```
Comment (ilija42) [inline core/capabilities/ccip/configs/evm/contract_reader.go:143]: Makram confirmed that this is not needed

File: core/capabilities/ccip/configs/evm/contract_reader.go [hunk 3]
```diff
@@ -140,11 +140,6 @@ var SourceReaderConfig = evmrelaytypes.ChainReaderConfig{
...
-						GenericDataWordNames: map[string]uint8{
```
Comment (ilija42) [inline core/capabilities/ccip/configs/evm/contract_reader.go:198]: Checked with Makram, this is not needed anymore

**Review by EasterTheBunny — DISMISSED** [review]

**Review by Farber98 — DISMISSED** [review]

File: contracts/src/v0.8/shared/test/helpers/ChainReaderTester.sol [hunk 4]
```diff
@@ -26,11 +26,11 @@ struct InnerTestStruct {
 contract ChainReaderTester {
   event Triggered(
     int32 indexed field,
-    string differentField,
     uint8 oracleId,
     uint8[32] oracleIds,
     address Account,
     address[] Accounts,
+    string differentField,
     int192 bigField,
```
Comment (silaslenihan) [inline contracts/src/v0.8/shared/test/helpers/ChainReaderTester.sol:34]: I don't think it is necessary to move the `string` field, rather we should be able to just use index 1 for `oracleId` in the `ChainReader` config and the LogPoller will read the 2nd 32 byte data word. But maybe that will break things that are outside the scope for this ticket that I will handle in my PR.

Comment (ilija42) [reply contracts/src/v0.8/shared/test/helpers/ChainReaderTester.sol:34]: The indexes from the cfg are not used in this way, which is why I just removed them. Indexes are calculated internally on CR init.

Comment (silaslenihan) [reply contracts/src/v0.8/shared/test/helpers/ChainReaderTester.sol:34]: I see, currently the loop in `getEventTypes` skips the rest of the fields if it has seen a dynamic type.

**Review by Farber98 — DISMISSED** [review]

**Review by EasterTheBunny — DISMISSED** [review]

**Review by silaslenihan — DISMISSED** [review]

**Review by silaslenihan — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**